src/main/kotlin/org/rust/lang/core/psi/RsFile.kt

```diff
@@ -70,50 +71,134 @@ class RsFile(
-    override val typeAliasList: List<RsTypeAlias> get() = findItems(TYPE_ALIAS)
-    override val useItemList: List<RsUseItem> get() = findItems(USE_ITEM)
-    override val modDeclItemList: List<RsModDeclItem> get() = findItems(MOD_DECL_ITEM)
-    override val externCrateItemList: List<RsExternCrateItem> get() = findItems(EXTERN_CRATE_ITEM)
-    override val foreignModItemList: List<RsForeignModItem> get() = findItems(FOREIGN_MOD_ITEM)
-    override val macroDefinitionList: List<RsMacroDefinition> get() = findItems(MACRO_DEFINITION)
-
-    private inline fun <reified T : RsCompositeElement> findItems(elementType: IElementType): List<T> {
-        val stub = stub
-        return if (stub != null) {
-            @Suppress("UNCHECKED_CAST")
-            stub.getChildrenByType(elementType, { arrayOfNulls<T>(it) })
-                .asList() as List<T>
-        } else {
-            PsiTreeUtil.getChildrenOfTypeAsList(this, T::class.java)
-        }
+    override val functionList: List<RsFunction> get() = itemsCache.functionList
+    override val modItemList: List<RsModItem> get() = itemsCache.modItemList
+    override val constantList: List<RsConstant> get() = itemsCache.constantList
+    override val structItemList: List<RsStructItem> get() = itemsCache.structItemList
+    override val enumItemList: List<RsEnumItem> get() = itemsCache.enumItemList
+    override val implItemList: List<RsImplItem> get() = itemsCache.implItemList
+    override val traitItemList: List<RsTraitItem> get() = itemsCache.traitItemList
+    override val typeAliasList: List<RsTypeAlias> get() = itemsCache.typeAliasList
+    override val useItemList: List<RsUseItem> get() = itemsCache.useItemList
+    override val modDeclItemList: List<RsModDeclItem> get() = itemsCache.modDeclItemList
+    override val externCrateItemList: List<RsExternCrateItem> get() = itemsCache.externCrateItemList
+    override val foreignModItemList: List<RsForeignModItem> get() = itemsCache.foreignModItemList
+    override val macroDefinitionList: List<RsMacroDefinition> get() = itemsCache.macroDefinitionList
+
+    private class ItemsCache(
+        val functionList: List<RsFunction>,
+        val modItemList: List<RsModItem>,
+        val constantList: List<RsConstant>,
+        val structItemList: List<RsStructItem>,
+        val enumItemList: List<RsEnumItem>,
+        val implItemList: List<RsImplItem>,
+        val traitItemList: List<RsTraitItem>,
+        val typeAliasList: List<RsTypeAlias>,
+        val useItemList: List<RsUseItem>,
+        val modDeclItemList: List<RsModDeclItem>,
+        val externCrateItemList: List<RsExternCrateItem>,
+        val foreignModItemList: List<RsForeignModItem>,
+        val macroDefinitionList: List<RsMacroDefinition>
+    )
+
+    @Volatile
+    private var _itemsCache: ItemsCache? = null
+
+    override fun subtreeChanged() {
+        super.subtreeChanged()
+        _itemsCache = null
     }
+
+    private val itemsCache: ItemsCache
```

Comment:
May be it's better to use `by lazy(PUBLICATION)` instead?

Reply:
It won't work unfortunately because we need to clear cache manually in `subtreeChanged`.